time-series-project/tsp_solution.py

```diff
@@ -0,0 +1,255 @@
+
+# %%
+plot_acf(rl_df_diff1.dropna(), lags = np.arange(0, 600, 7)); # q = 1 # look at first 600 lags, every 7
+plt.ylim(-0.1,0.1)
+
+# %%
+plot_pacf(rl_df_diff1.dropna()); # p = 1
+
+# Do ARIMA if data has no seasonality
+
+# %%
+rl_group = [rl_df.iloc[i::7] for i in range(7)] # 7 because it's 7 hours each day
+rl_group
+
+# %%
+kruskal(*rl_group) # p-value is basically 1, so there's insufficient evidence for seasonality.
+
+# %%
+# Instead of getting hourly data, you might want to get minutely 
+# 60 mins * number of days for each month * 7 hours
+
+# %%
+p = list(range(1,4))
+d = list(range(0,3))
+q = list(range(0,3))
+
+pdq_list = list(itertools.product(p,d,q))
+
+# %%
+len(rl_df)
+
+# %%
+len(rl_df) * 0.75
+
+# %%
+len(rl_df) - 1317 #439
+
+# %%
+0.75 * len(rl_df)
+
+# %%
+train = rl_df[:1317]
+test = rl_df[1317:]
+rmse = []
+
+# %% [markdown]
+# # From the original data, store it into a list, and x-values to np.linspace
+
+# %%
+for i in tqdm(range(len(pdq_list))):
+    m = model.ARIMA(train, order = pdq_list[i])
+    fitted = m.fit()
+    preds = fitted.forecast(439)
+    rmse.append(mean_squared_error(test, preds, squared = False))
+
+# %%
+np.argmin(rmse) # 20
+
+# %%
+rmse[20]
+
+# %%
+pdq_list[20] 
+
+# %%
+best_model = model.ARIMA(rl_df, order = pdq_list[20])
+results = best_model.fit()
+
+# %%
+1318 + 439
+
+# %%
+1758 + 439
+
+# %%
+# We have to convert the timestamps to numerical integers since Python's having problems with determining 
+# the future timestamps beyond today, for some reason. This is also to make plotting the data possible
+
+ls1 = np.arange(1,1757, 1) # Timestamps of the original data converted to a series of subsequent integers (1757 is non-inclusive)
+ls2 = np.arange(1757, 2196, 1) # Would-be timestamps of the forecasted values
+original_vals = rl_df['Open'].values # Original values stored separately
+
+# %%
+predicted_vals = results.fittedvalues.values
+
+# I did .values at the end since I want to exclude the timestamps, otherwise
+# we run into an issue of trying to plot out the values
+
+# %%
+flist = results.forecast(439).values
+
+# %%
+len(ls1)
+
+# %%
+len(ls2)
+
+# %%
+len(original_vals)
+
+# %%
+len(flist)
+
+# %%
+plt.plot(ls1, original_vals, color = 'aqua')
+plt.plot(ls1, predicted_vals, color = 'red')
+plt.plot(ls2, flist, color = 'navy')
+
+# %%
+# That particular frequency (peak) has a higher magnitude, so that frequency contributes significantly
+
+# %%
+# plt.plot(rl_df, color = 'blue')
+# plt.plot(results.fittedvalues, color = 'yellow')
+
+# %%
+# I decided to plot a periodogram in order to further investigate any seasonality in the data
+f, Pxx_den = periodogram(rl_df['Open']) # Anything above two times above the original frequency
+
+# %%
+plt.figure()
+plt.semilogy(f[1:], Pxx_den[1:]) # Semi logarithmic y
+plt.xlabel('Frequency [Hz]')
+plt.ylabel('Power Spectral Density [V**2/Hz]')
+plt.title('Periodogram')
+plt.show
+
+# %%
+data_array = np.array([ls1, predicted_vals, original_vals])

```

Comment:
You could save the actual timestamps since this is only for the original data (not forecasts)